### PR TITLE
add a `hab pkg header` command to read the hart file header

### PR DESCRIPTION
**Pull request by smurawski**
I had a few minutes this morning and thought this might be a handy subcommand.

/cc @juliandunn 

Signed-off-by: Steven Murawski <steven.murawski@gmail.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/hab/src/cli.rs [hunk 1]
```diff
@@ -209,6 +209,13 @@ pub fn get() -> App<'static, 'static> {
                     "A path to a Habitat Artifact \
                     (ex: /home/acme-redis-3.0.7-21120102031201-x86_64-linux.hart)")
             )
+            (@subcommand header =>
+                (about: "Returns the Habitat Artifact header")
+                (aliases: &["hea", "head", "heade", "header"])
```
Comment (smurawski) [inline components/hab/src/cli.rs]: This command has fewer alias because of some lettering overlap with help.

**Comment by juliandunn** [comment]
Thanks! This could use a test I think - I imagine we _might_ need to change the header when we get to Windows, and I'm sure we'll break this if so.

**Comment by smurawski** [comment]
@juliandunn since I'm using our artifact api and not manually parsing the header, I'm using tested code paths which will need to be updated if we change the package header.  The subcommand code is mostly output using our ui library functions, so there isn't much to independently test.

**Comment by smurawski** [comment]
hab_pkg_header

**Comment by adamhjk** [comment]
I would make the output here more parse-able - it's pretty this way, but it'll not be very useful if you turn the colors off. Should it be colon-separated? @fnichol?

**Comment by smurawski** [comment]
@adamhjk how about a switch to output JSON?

**Comment by adamhjk** [comment]
@smurawski your powershell is showing. :) a good example of how you might use this, to say make a choice based on the type of hart you have:

```bash
$ hab pkg header foo.hart | grep "Format Version" | cut -d ":" -f 1
```

Would print the version if the output had "Format Version: HART-1"

**Comment by fnichol** [comment]
We've been favoring the simplest structured output as we can get away with, primarily so that in minimal environments a user wouldn't need extra tools or languages (in other words, can we pass the "do you have cat-and-cut?" test). In that spirit the colon-delimited version looks to work decently. However, I can't think of any subcommands that (as of yet) have structured output meant to be consumed by other scripts, aside from `hab pkg path`.

With a product hat on, I'd be interested in what value we're providing with this command? Up to this point the structure of the package metadata/frontmatter has been internal implementation details that we could conceivably change in a forwards compatible way. Don't get me wrong, there is most likely value to this--I'm just coming up with few ideas this morning.

**Comment by smurawski** [comment]
The idea for the subcommand @fnichol was primarily for habitat developers as they build and validate packages (rather than having to remember `head -n5` or some such to return the package format (especially since on Windows that command would be different and, depending on your version of powershell, could be pretty ugly).

**Comment by ryankeairns** [comment]
@smurawski , et al, I'll post a quick mockup of slightly revised layout for the output.

**Comment by reset** [comment]
@smurawski @fnichol @ryankeairns I am inclined to agree with @fnichol here, the header is an implementation detail which should be invisible to the end-user. What if we made this a hidden CLI command?

**Comment by ryankeairns** [comment]
@reset that makes sense to me... you're suggesting it basically doesn't show up in the `help` list of commands, right?

**Comment by ryankeairns** [comment]
As for the output, I'd propose a format like this:

hab pkg header

**Comment by reset** [comment]
@ryankeairns that's right, it's still there but it's not showing up in the help

**Comment by smurawski** [comment]
@ryankeairns @reset @fnichol 
Closer?
hab_pkg_header_hidden

**Comment by reset** [comment]
@smurawski I like it!

**Comment by smurawski** [comment]
@reset  I don't have the nice green from @ryankeairns, but I think I can sort that out.

**Comment by ryankeairns** [comment]
@smurwaski looks great, thanks for making the changes.

File: components/hab/src/command/pkg.rs [hunk 2]
```diff
@@ -571,3 +571,33 @@ pub mod verify {
...
+    use std::path::Path;
+
+    use common::ui::UI;
+    use hcore::crypto::artifact;
+    use std::io::{self, Write};
+
+    use error::Result;
+
+    pub fn start(ui: &mut UI, src: &Path) -> Result<()> {
+        try!(ui.begin(format!("Reading Artifact Header for {}", &src.display())));
```
Comment (juliandunn) [inline components/hab/src/command/pkg.rs]: I would recommend not capitalizing all this, and also, we have in the past decided not to refer to `.hart`s as artifacts, just as packages. So altogether:

`Reading package header for {}`

File: components/hab/src/command/pkg.rs [hunk 3]
```diff
@@ -571,3 +571,33 @@ pub mod verify {
...
+        if artifact_header.is_ok() {
+            let header = artifact_header.unwrap();
+            try!(io::stdout()
+                .write(format!("Format Version : {}\n", header.format_version).as_bytes()));
+            try!(io::stdout().write(format!("Key Name       : {}\n", header.key_name).as_bytes()));
+            try!(io::stdout().write(format!("Hash Type      : {}\n", header.hash_type).as_bytes()));
+            try!(io::stdout()
+                .write(format!("Raw Signature  : {}\n", header.signature_raw).as_bytes()));
+        } else {
+            try!(ui.warn("Failed to Read Artifact Header."));
```
Comment (juliandunn) [inline components/hab/src/command/pkg.rs]: Ditto here:

`Failed to read package header`

File: components/hab/src/command/pkg.rs [hunk 4]
```diff
@@ -571,3 +571,33 @@ pub mod verify {
...
+        }
+        try!(ui.para(""));
+        try!(ui.end(format!("Finished Reading Artifact Header for {}.", &src.display())));
```
Comment (juliandunn) [inline components/hab/src/command/pkg.rs]: I'd suggest just eliminating this message.

File: components/hab/src/command/pkg.rs (hunk 2 above)
Comment (smurawski) [inline components/hab/src/command/pkg.rs]: `Artifact` is the term we use elsewhere in the cli help, so we should probably clean that up as well if we just want to use package. (I've got no problem with that, I was just trying to stay consistent with the rest of the cli.

**Comment by smurawski** [comment]
@juliandunn - updated messaging per your comments
image

**Comment by juliandunn** [comment]
LGTM, let's get a 👍 from one of the maintainers and perhaps we can :shipit: in tomorrow's release!

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/170925828).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/170925828).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**